targets/TARGET_NUVOTON/TARGET_NANO100/us_ticker.c

```diff
@@ -78,60 +70,109 @@ void us_ticker_init(void)
-        
-    do {
-        uint32_t major_minor_us;
-        uint32_t minor_us;
-
-        // NOTE: As TIMER_DR = TIMER_CMPR and counter_major has increased by one, TIMER_DR doesn't change to 0 for one tick time.
-        // NOTE: As TIMER_DR = TIMER_CMPR or TIMER_DR = 0, counter_major (ISR) may not sync with TIMER_DR. So skip and fetch stable one at the cost of 1 clock delay on this read.
-        do {
-            core_util_critical_section_enter();
-            
-            // NOTE: Order of reading minor_us/carry here is significant.
-            minor_us = TIMER_GetCounter(timer0_base) * US_PER_TMR0HIRES_CLK;
-            uint32_t carry = (timer0_base->ISR & TIMER_ISR_TMR_IS_Msk) ? 1 : 0;
-            // When TIMER_DR approaches TIMER_CMPR and will wrap soon, we may get carry but TIMER_DR not wrapped. Hanlde carefully carry == 1 && TIMER_DR is near TIMER_CMPR.
-            if (carry && minor_us > (US_PER_TMR0HIRES_INT / 2)) {
-                major_minor_us = (counter_major + 1) * US_PER_TMR0HIRES_INT;
-            }
-            else {
-                major_minor_us = (counter_major + carry) * US_PER_TMR0HIRES_INT + minor_us;
-            }
-            
-            core_util_critical_section_exit();
-        }
-        while (minor_us == 0 || minor_us == US_PER_TMR0HIRES_INT);
-        
-        return (major_minor_us / US_PER_TICK);
+
+    ticker_last_read_clk = TIMER_GetCounter(timer0_base);
+    return  (ticker_last_read_clk / NU_TMRCLK_PER_TICK);
+}
+
+void us_ticker_set_interrupt(timestamp_t timestamp)
+{
+    TIMER_Stop((TIMER_T *) NU_MODBASE(timer1hires_modinit.modname));
+
+    /* We need to get alarm interval from alarm timestamp `timestamp` to configure H/W timer.
+     *  
+     * Because both `timestamp` and xx_ticker_read() would wrap around, we have difficulties in distinguishing
+     * long future event and past event. To distinguish them, we need `tick_last_read` against which 
+     * `timestamp` is calculated out. In timeline, we would always have below after fixing wrap-around:
+     * (1) tick_last_read <= present_clk
+     * (2) tick_last_read <= alarm_ts_clk
+     *
+     *
+     * 1. Future event case:
+     *
+     * tick_last_read     present_clk                         alarm_ts_clk
+     * |                  |                                   |
+     * --------------------------------------------------------
+     * |-alarm_intvl1_clk-|
+     * |-------------------alarm_intvl2_clk-------------------|
+     *
+     * 2. Past event case:
+     *
+     * tick_last_read     alarm_ts_clk                        present_clk
+     * |                  |                                   |
+     * --------------------------------------------------------
+     * |-------------------alarm_intvl1_clk-------------------|
+     * |-alarm_intvl2_clk-|
+     *
+     * Unfortunately, `tick_last_read` is not passed along the xx_ticker_set_interrupt() call. To solve it, we
```

Comment:
shouldnt this work without storing the last read? There should not be any event in the past when set_interrupt is called (processed via different branch in the upper layer) ?

Reply:
@0xc0170 Assume at `t1`, upper layer calls `us_ticker_read` and has handled all past events; at `t2`, upper layer calls `us_ticker_set_interrupt` to schedule future event (`timestamp`) relative to `t1`. There is very low possibility that future event (`timestamp`) relative to `t1` becomes past event (`timestamp`) relative to `t2`. But for robustness, I need the last read (`t1`) as a reference point to distinguish the order of `t1`, `t2`, and `timestamp` in timeline and then to distinguish `timestamp` is a future or past event.